### PR TITLE
Issue #1624 and #1653 - date range search improvements

### PR DESCRIPTION
**Pull request by michaelwschroeder**
Signed-off-by: Mike Schroeder <mschroed@us.ibm.com>

### CODE REVIEW

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java [hunk 1]
```diff
@@ -1055,7 +1056,32 @@ protected WhereFragment getQuantityFilter(QueryParameter queryParm, String param
...
-        util.executeBehavior(where, queryParm, paramAlias);
+        
+        // It is possible that multiple date parameters could be chained together if there were
+        // multiple query parameters specified for the same date search parameter and we were
+        // able to consolidate them. Check specifically if we have a consolidated parameter which
+        // specifies a range, and if so, build a range filter.
+        if (queryParm.isChained() && queryParm.getChain().size() == 1) {
+            Prefix prefix1 = queryParm.getValues().get(0).getPrefix();
+            Prefix prefix2 = queryParm.getNextParameter().getValues().get(0).getPrefix();
+            if ((Prefix.GT.equals(prefix1) || Prefix.GE.equals(prefix1) || Prefix.SA.equals(prefix1)) &&
```
Comment (tbieste) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java:1067]: Could prefix1 have the LT/LE/EB and prefix2 have the GT/GE/SA, or is that permutation not possible?

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/NewQueryBuilder.java [hunk 2]
```diff
@@ -598,4 +610,177 @@ private void addResourceTypeExtension(SearchQuery domainModel, List<String> reso
...
+                Instant ebBound = null;
+                QueryParameter gteBoundParm = null;
+                QueryParameter lteBoundParm = null;
+                QueryParameter saBoundParm = null;
+                QueryParameter ebBoundParm = null;
+                for (QueryParameter queryParm : queryParameters) {
+                    QueryParameterValue queryParmValue = queryParm.getValues().get(0);
+                    Prefix prefix = queryParmValue.getPrefix();
+                    Instant valueLowerBound = queryParmValue.getValueDateLowerBound();
+                    Instant valueUpperBound = queryParmValue.getValueDateLowerBound();
```
Comment (tbieste) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/NewQueryBuilder.java:686]: Should this be "getValueDateUpperBound()"?

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java (hunk 1 above)
Comment (tbieste) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java:1067]: Ah, ok. The consolidateDateParms() method explicitly puts them in that order.  Might be good to handle either order so the two methods are not assuming order; or at least document in consolidateDateParms() that getDateFilter assumes the order.

Comment (michaelwschroeder) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java]: updated to handle either order

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/NewQueryBuilder.java (hunk 2 above)
Comment (michaelwschroeder) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/NewQueryBuilder.java]: yes, updated - nice catch

**Review by tbieste — APPROVED** [review]
LGTM

**Pull request merged**